packages/optimizer/lib/transformers/SeparateKeyframes.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const css = require('css');
+const stringifyOptions = {indent: 0, compress: true};
+/**
+ * SeparateKeyframes - moves keyframes, media, and support from amp-custom
+ * to amp-keyframes.
+ */
+class SeparateKeyframes {
+  transform(tree) {
+    const html = tree.root.firstChildByTag('html');
+    if (!html) return;
+    const head = html.firstChildByTag('head');
+    if (!head) return;
+    const body = html.firstChildByTag('body') || head;
+    let stylesCustomTag;
+    let stylesKeyframesTag;
+    const headTags = head.children;
+
+    for (let i = 0; i < headTags.length; i++) {
+      const tag = headTags[i];
+      if (tag.tagName !== 'style') continue;
+
+      if (!stylesKeyframesTag && tag.hasAttribute('amp-keyframes')) {
+        stylesKeyframesTag = tag;
+      }
+      if (!stylesCustomTag && tag.hasAttribute('amp-custom')) {
+        stylesCustomTag = tag;
+      }
+    }
+
+    // If no custom styles, there's nothing to do
+    if (!stylesCustomTag) return;
+    let stylesText = stylesCustomTag.children[0];
+
+    if (!stylesText || !stylesText.data) return;
+    stylesText = stylesText.data;
+
+    const cssTree = css.parse(stylesText);
+    const keyframesTree = {
+      type: 'stylesheet',
+      stylesheet: {
+        rules: [],
+      },
+    };
+
+    cssTree.stylesheet.rules = cssTree.stylesheet.rules.filter((rule) => {
+      const doMove = (
+        rule.type === 'media' ||
+        rule.type === 'supports' ||
+        rule.type === 'keyframes'
+      );
+      if (doMove) {
+        keyframesTree.stylesheet.rules.push(rule);
+      }
+      return !doMove;
+    });
+
+    // if no rules moved nothing to do
+    if (!keyframesTree.stylesheet.rules.length) return;
+    let hadKeyframesTag = true;
+
+    if (!stylesKeyframesTag) {
+      stylesKeyframesTag = body.firstChildByTag('style');
```

Comment:
Do we need to search again for the `style` tag? I thought we identified all style tags at the beginning.

Reply:
At the beginning I'm checking the head for the `style[amp-custom]` and also checking to see if there's a `style[amp-keyframes]` tag. 

Here I'm checking if a `style[amp-keyframes]` tag wasn't found in the head then I'm making sure there isn't one in the body

Reply:
Ah got, sorry missed that.